spmd/__init__.py

```diff
@@ -1,106 +1,96 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates
-from typing import Sequence, Optional, cast
+from typing import Optional, Callable, Tuple, Union
 import torch
 import torch.nn as nn
-from spmd.tensor import DTensor, Placement, Shard, Replicate
+from spmd.tensor import distribute_tensor
+from spmd.tensor.api import DTensor
 from spmd.tensor.device_mesh import get_global_device_mesh, DeviceMesh
+from spmd.tensor.placement_types import Replicate, Shard
 
-torch.__future__.set_overwrite_module_params_on_conversion(True)
 
-
-def distribute_tensor(
-    tensor: torch.Tensor,
+def distribute_module(
+    module: nn.Module,
     device_mesh: Optional[DeviceMesh] = None,
-    placements: Optional[Sequence[Placement]] = None,
-) -> DTensor:
+    partition_fn: Optional[Callable[[str, nn.Module], None]] = None,
+    input_fn: Optional[Callable[..., None]] = None,
+    output_fn: Optional[Callable[..., None]] = None,
+) -> nn.Module:
     """
-    Distribute a torch.Tensor to the `device_mesh` according to the `placements`
-    specified. The rank of `device_mesh` and `placements` must be the same.
-
+    This function converts all module parameters to :class:`DTensor` parameters
+    according to the `partition_fn` specified. It could also control the input or
+    output of the module by specifying the `input_fn` and `output_fn`. (i.e. convert
+    the input to :class:`DTensor`, convert the output back to torch.Tensor)
     Args:
-        tensor (torch.Tensor): torch.Tensor to be distributed
-        device_mesh (:class:`DeviceMesh`, optional): DeviceMesh to distribute the
-            tensor, if not specified, must be called under a DeviceMesh context
-            manager, default: None
-        placements (List[:class:`Placement`], optional): the placements that
-            describes how to place the tensor on DeviceMesh, must have the same
-            number of elements as `device_mesh.ndim`. If not specified, we will
-            by default replicate the tensor across the `device_mesh` from the
-            first rank of each dimension of the `device_mesh`.
+        module (:class:`nn.Module`): user module to be partitioned.
+        device_mesh (:class:`DeviceMesh`): the device mesh to place the module.
+        partition_fn (Callable): the function to partition parameters (i.e. shard certain
+            parameters across the `device_mesh`). If `partition_fn` is not specified,
+            by default we replicate all module parameters of `module` across the mesh.
+        input_fn (Callable): specify the input distribution, i.e. could control how the
+            input of the module is sharded. `input_fn` will be installed as a module
+            forward_pre_hook.
+        output_fn (Callable): specify the output distribution, i.e. could control how the
+            output is sharded, or convert it back to torch.Tensor. output_fn will be
+            installed as a module forward_hook.
```

Comment:
Should this be post forward hook?

Reply:
I guess its call forward_hook but was trying to see if we could clarify when it ran.

Reply:
yeah sure I'll update the comment to clarify